include/deal.II/sundials/kinsol.h

```diff
@@ -423,29 +431,29 @@ namespace SUNDIALS
      * SolutionStrategy::newton, SolutionStrategy::linesearch, or
      * SolutionStrategy::picard strategies were selected.
      *
-     * This function should return:
-     * - 0: Success
-     * - >0: Recoverable error (KINSOL will try to change its internal
-     * parameters and attempt a new solution step)
-     * - <0: Unrecoverable error the computation will be aborted and an
-     * assertion will be thrown.
+     * @note This variable represents a
+     * @ref GlossUserProvidedCallBack "user provided callback".
+     * See there for a description of how to deal with errors and other
+     * requirements and conventions. In particular, KINSOL can deal
+     * with "recoverable" errors in some circumstances, so callbacks
+     * can throw exceptions of type RecoverableUserCallbackError.
      */
-    std::function<int(const VectorType &src, VectorType &dst)> residual;
+    std::function<void(const VectorType &src, VectorType &dst)> residual;
```

Comment:
Is this really backwards-compatible? I don't think you can assign existing  callbacks that look like 

```c++
[](const VectorType &src, VectorType &dst) { 
  // some code
  return 0;
}
```
due to the (deduced) return type being int instead of void.

Reply:
Nvm, my quick check in compiler explorer which seemed to indicate this is not possible was flawed for another reason... I can confirm this works ;)

Reply:
:-) `std::function` really does not care very much what the object is that it gets to represent, as long as it can be called with the arguments given when one calls `std::function::operator()`, and as long as the result of the called thing can be assigned to the return code of the `std::function` object (which I changed from `int` to `void` -- anything can be assigned to `void`, or put differently: the return value of the thing the `std::function` stores is simply ignored).

All of this means that this code snippet compiles and runs just fine:
```
#include <functional>

int f(int);

int main ()
{
  std::function<void(double)> fun = &f;    // note mismatch in return type and argument
  fun(1.234);   // ok -- fun.operator() takes a 'double' as argument, which is implicitly converted to 'int' when calling 'f'
}
```

Reply:
Independent of if it compiles or no not, we probably need a change-log entry, since if the return value is just ignored this means that means the behavior will be different.